feat: do NOT add delete entries in commit log when expired keys are deleted

srieteja opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"
Please provide the following information:
-->
closes https://github.com/atsign-foundation/at_server/issues/1660
**- What I did**
- Do not add CommitEntry for Delete Op on expired keys (disabled for now, will be enabled later)
- Introduced an optional param in scheduleKeyExpireTask() called runInterval which takes Duration and schedules the expiryTask at this timeInterval. (This is an improvement as previously this method would accept only time in mins and in type String)

**- How I did it**
- Introduce an optional named param in deleteExpiredKeys() called skipCommit, this decides whether a commit should be skipped when an expired key is deleted. [defaults to false; meaning commit will be created]
- Based on the method param above, set the skipCommit flag while calling remove() from deleteExpiredKeys() [in hiveKeyStore and notificationKeyStore]

**- How to verify it**
- new unit tests added should verify the behaviour of expired keys w.r.t. the changes

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
feat: do NOT add delete entries in commit log when expired keys are deleted

srieteja committed:
fix: do NOT add DELETE entries for keys and notifications to commitLog

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
test: add unit tests to verify changes

srieteja committed:
test: fix tests

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
build: update changelog and pub version to 3.0.60

srieteja committed:
Merge remote-tracking branch 'origin/delete_commit_nosync' into delete_commit_nosync

srieteja committed:
fix: remove unnecessary code

srieteja committed:
test: reduce expired check test delay to 23s

srieteja committed:
feat: optionally allow Duration as method param for scheduleKeyExpireTask()

gkc commented:
@srieteja We have a chicken and egg scenario here; until we can be sure that most apps are using the latest AtClient, those clients will still be dependent on the commit log and sync responses to trigger deletion on the client side.

Therefore for now we need to
- add a flag to the atServer's config for this feature, defaulted to false (not enabled) for now
- and add a way of setting that flag dynamically for testing purposes (add it to `ModifiableConfigs`, add a subscriber somewhere, etc etc)
- and modify tests so they verify both the current behaviour with the flag not enabled, and the new behaviour with the flag enabled
- and add a test to assert that the flag is by default not enabled

At some time in the future, we can change the config to enable this feature by default. I'd suggest create a ticket for that and reference that new ticket in the description of #1660 and we'll keep assigning the ticket to the "sprint after next" until we think we're ready to go.

Make sense? cc @murali-shris @sitaram-kalluri

gkc reviewed (changes requested):
See comment I left

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

sitaram-kalluri commented on packages/at_persistence_secondary_server/CHANGELOG.md:
```diff
@@ -1,4 +1,5 @@
 ## 3.0.60
+- feat: delete entries for expired keys are not committed to the commitLog [feature not enabled yet]
```
FYI, The package with 3.0.60 is published. Please upgrade the version

srieteja replied:
addressed in [b1c9493](https://github.com/atsign-foundation/at_server/pull/1713/commits/b1c949376913a6f090efed26cf6099ce7bde78bd)

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
feat: introduce flag to optimize commits for expired keys

srieteja committed:
docs: update changelog

srieteja commented:
Added flags in hive_manages/scheduleKeyExpiryTask() and hive_keystore/deleteExpiredKeys() to enable the feature being written in this PR. Also introduced config in AtSecondaryServer/AtSecondaryConfig and passed the same to the hive_manager/keystore to enable the feature in the PR (see https://github.com/atsign-foundation/at_server/pull/1727)

srieteja committed:
tests: more tests

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
hive_manager: remove incorrect schedule snippet

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
reformat: run dart formatter

srieteja committed:
Merge remote-tracking branch 'origin/delete_commit_nosync' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja commented:
I have tested the functionality and how this feature interacts with the client. @purnimavenkatasubbu has tested this along with https://github.com/atsign-foundation/at_server/pull/1727 and https://github.com/atsign-foundation/at_client_sdk/pull/1187 with buzz and wavi. Everything seems to be working fine, Ready to be reviewed

srieteja committed:
build: update version in pubspec to 3.0.61

srieteja committed:
Merge remote-tracking branch 'origin/delete_commit_nosync' into delete_commit_nosync

gkc commented:
@srieteja in the PR description you say that you 
> - set the skipCommit flag to true in deleteExpiredKeys() [in hiveKeyStore and notificationKeyStore]

that is only true for notificationKeyStore. In hiveKeyStore deleteExpiredKeys takes an "optimizeCommits" flag

gkc commented on packages/at_persistence_secondary_server/lib/src/keystore/hive_keystore.dart:
```diff
@@ -341,7 +340,9 @@ class HiveKeystore implements SecondaryKeyStore<String, AtData?, AtMetaData?> {
   @server
-  Future<bool> deleteExpiredKeys() async {
+  @client
+  Future<bool> deleteExpiredKeys({bool? optimizeCommits = false}) async {
```
Why not call it 'skipCommit' same as the existing flag name?

srieteja replied:
makes more sense, let me get that changed

srieteja replied:
change made as part of [fefa7d4](https://github.com/atsign-foundation/at_server/pull/1713/commits/fefa7d41067053944799aff620786d1a84b19694)

gkc committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
refactor: rename optimizeCommits -> skipCommits

srieteja committed:
Merge remote-tracking branch 'origin/delete_commit_nosync' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

srieteja committed:
tests: rename optimizeCommits -> skipCommits

srieteja committed:
Merge remote-tracking branch 'origin/delete_commit_nosync' into delete_commit_nosync

srieteja committed:
Merge branch 'trunk' into delete_commit_nosync

gkc reviewed (approved):
LGTM. Nice work on the tests

gkc commented:
> @srieteja in the PR description you say that you
> 
> > * set the skipCommit flag to true in deleteExpiredKeys() [in hiveKeyStore and notificationKeyStore]
> 
> that is only true for notificationKeyStore. In hiveKeyStore deleteExpiredKeys takes an "optimizeCommits" flag

@srieteja Please can you correct the PR description so it is accurate

srieteja commented:
Updated the PR description @gkc
